File: src/lua_zbench.zig
//! lua-zbench - Lua microbenchmarking framework
//! Core module that bridges Zig, zBench, and Lua via zlua.
//!
//! Phase 1: Basic binding and timing verification.
//! Provides `lua_high_res_time` function that returns high-resolution
//! timestamps as Lua numbers (seconds as f64).

const std = @import("std");
const zlua = @import("zlua");
const zbench = @import("zbench");

const Lua = zlua.Lua;

/// High-resolution time function for Lua.
/// Returns the current time in seconds as a f64 (nanosecond precision).
/// Usage in Lua:
///   local time = lua_zbench.gettime()
fn luaHighResTime(lua: *Lua) callconv(.c) c_int {
    // Use POSIX clock_gettime for reliable high-resolution timing
    var ts: std.c.timespec = undefined;
    _ = std.c.clock_gettime(std.c.CLOCK.REALTIME, &ts);
    const seconds = @as(f64, @floatFromInt(ts.sec)) +
        @as(f64, @floatFromInt(ts.nsec)) / 1_000_000_000.0;
    lua.pushNumber(seconds);
    return 1;
}

// Thread-local storage for passing context to benchmark functions
threadlocal var bench_lua: ?*Lua = null;
threadlocal var bench_ref: c_int = 0;

fn benchFuncWrapper(_: std.mem.Allocator) void {
    if (bench_lua) |lua| {
        _ = callLuaFunc(lua, bench_ref);
    }
}

/// Run a benchmark from Lua.
/// Usage in Lua:
///   local results = lua_zbench.run(name, func, opts)
/// Returns a table with: name, iterations, mean_ns, min_ns, max_ns,
///   stddev_ns, p75_ns, p99_ns, p99_9_ns, total_ns
fn luaBenchRun(lua: *Lua) callconv(.c) c_int {
    const name = lua.checkString(1);
    if (!lua.isFunction(2)) {
        _ = lua.pushString("argument #2 must be a function");
        return 0;
    }

    var config = zbench.Config{};
    if (lua.getTop() >= 3 and lua.isTable(3)) {
        _ = lua.getField(3, "time_budget_ms");
        if (!lua.isNil(-1)) {
            config.time_budget_ns = @intFromFloat((lua.toNumber(-1) catch 0.0) * 1_000_000.0);
        }
        lua.pop(1);

        _ = lua.getField(3, "max_iterations");
        if (!lua.isNil(-1)) {
            config.max_iterations = @intFromFloat(lua.toNumber(-1) catch 0.0);
        }
        lua.pop(1);

        _ = lua.getField(3, "track_memory");
        if (!lua.isNil(-1)) {
            config.track_allocations = lua.toBoolean(-1);
        }
        lua.pop(1);
    }

    // Push the function and get a reference to it
    lua.pushValue(2);
    const func_ref = lua.ref(zlua.registry_index);

    const allocator = std.heap.page_allocator;

    // Adaptive batch size optimization:
    // First, do a quick probe run to estimate the function's execution time.
    // If single call takes > 10µs, reduce max_iterations to avoid over-batching.
    {
        var threaded: std.Io.Threaded = .init_single_threaded;
        const io = threaded.io();
        const probe_start = std.Io.Timestamp.now(io, .awake).nanoseconds;
        bench_lua = lua;
        bench_ref = func_ref;
        benchFuncWrapper(allocator);
        const probe_elapsed = std.Io.Timestamp.now(io, .awake).nanoseconds - probe_start;
        bench_lua = null;
        bench_ref = 0;

        // If probe shows > 10µs per call, reduce iterations to avoid over-batching
        if (probe_elapsed > 10_000) {
            // For slow functions, we don't need many iterations
            if (config.max_iterations > 100) {
                config.max_iterations = 100;
            }
        }
    }

    // Create a benchmark and add the Lua function as a benchmark
    var bench = zbench.Benchmark.init(allocator, config);

    // Set thread-local context before adding benchmark
    bench_lua = lua;
    bench_ref = func_ref;

    bench.add(name, benchFuncWrapper, .{}) catch {
        bench_lua = null;
        bench_ref = 0;
        lua.unref(zlua.registry_index, func_ref);
        _ = lua.pushString("failed to add benchmark");
        return 0;
    };

    // Run the benchmark using iterator API to avoid stdout output
    // and collect results programmatically
    var threaded: std.Io.Threaded = .init_single_threaded;
    const io = threaded.io();

    var iter = bench.iterator() catch {
        bench_lua = null;
        bench_ref = 0;
        lua.unref(zlua.registry_index, func_ref);
        bench.deinit();
        _ = lua.pushString("failed to create iterator");
        return 0;
    };

    var result: ?zbench.Result = null;
    while (iter.next(io) catch {
        bench_lua = null;
        bench_ref = 0;
        lua.unref(zlua.registry_index, func_ref);
        bench.deinit();
        _ = lua.pushString("benchmark iteration failed");
        return 0;
    }) |step| {
        switch (step) {
            .progress => {},
            .result => |r| {
                result = r;
            },
        }
    }

    bench_lua = null;
    bench_ref = 0;
    lua.unref(zlua.registry_index, func_ref);

    const bench_result = result orelse {
        bench.deinit();
        _ = lua.pushString("no benchmark result");
        return 0;
    };

    // Calculate statistics from the readings
    const timings_ns = bench_result.readings.timings_ns;
    const stats = zbench.statistics.Statistics(u64).init(timings_ns) catch {
        bench_result.deinit();
        bench.deinit();
        _ = lua.pushString("failed to compute statistics");
        return 0;
    };

    // Build results table for Lua
    lua.newTable();

    // name
    _ = lua.pushString("name");
    _ = lua.pushString(bench_result.name);
    lua.setTable(-3);

    // iterations
    _ = lua.pushString("iterations");
    lua.pushNumber(@as(f64, @floatFromInt(bench_result.readings.iterations)));
    lua.setTable(-3);

    // total_ns
    _ = lua.pushString("total_ns");
    lua.pushNumber(@as(f64, @floatFromInt(stats.total)));
    lua.setTable(-3);

    // mean_ns
    _ = lua.pushString("mean_ns");
    lua.pushNumber(@as(f64, @floatFromInt(stats.mean)));
    lua.setTable(-3);

    // stddev_ns
    _ = lua.pushString("stddev_ns");
    lua.pushNumber(@as(f64, @floatFromInt(stats.stddev)));
    lua.setTable(-3);

    // min_ns
    _ = lua.pushString("min_ns");
    lua.pushNumber(@as(f64, @floatFromInt(stats.min)));
    lua.setTable(-3);

    // max_ns
    _ = lua.pushString("max_ns");
    lua.pushNumber(@as(f64, @floatFromInt(stats.max)));
    lua.setTable(-3);

    // p75_ns
    _ = lua.pushString("p75_ns");
    lua.pushNumber(@as(f64, @floatFromInt(stats.percentiles.p75)));
    lua.setTable(-3);

    // p99_ns
    _ = lua.pushString("p99_ns");
    lua.pushNumber(@as(f64, @floatFromInt(stats.percentiles.p99)));
    lua.setTable(-3);

    // p99_9_ns (zBench provides p995, we use it as p99.9 approximation)
    _ = lua.pushString("p99_9_ns");
    lua.pushNumber(@as(f64, @floatFromInt(stats.percentiles.p995)));
    lua.setTable(-3);

    // Memory tracking results (if enabled)
    if (bench_result.readings.allocations) |allocs| {
        const mem_stats = zbench.statistics.Statistics(usize).init(allocs.maxes) catch {
            bench_result.deinit();
            bench.deinit();
            _ = lua.pushString("failed to compute memory statistics");
            return 0;
        };

        _ = lua.pushString("alloc_max_bytes");
        lua.pushNumber(@as(f64, @floatFromInt(mem_stats.max)));
        lua.setTable(-3);

        _ = lua.pushString("alloc_mean_bytes");
        lua.pushNumber(@as(f64, @floatFromInt(mem_stats.mean)));
        lua.setTable(-3);

        _ = lua.pushString("alloc_count");
        lua.pushNumber(@as(f64, @floatFromInt(allocs.counts[0])));
        lua.setTable(-3);
    }

    bench_result.deinit();
    bench.deinit();

    return 1;
}

fn callLuaFunc(lua: *Lua, func_ref: c_int) c_int {
    // Use rawgeti via the C API directly
    const c_api = @as(*zlua.LuaState, @ptrCast(lua));
    _ = zlua.c.lua_rawgeti(c_api, zlua.registry_index, func_ref);
    lua.protectedCall(.{ .args = 0, .results = 0 }) catch {
        return -1;
    };
    return 0;
}

/// Run an empty baseline benchmark to measure measurement overhead.
/// Returns the same result format as luaBenchRun, representing the noise floor.
/// Usage in Lua:
///   local noise = lua_zbench.baseline(opts)
fn luaBaseline(lua: *Lua) callconv(.c) c_int {
    var config = zbench.Config{};
    if (lua.getTop() >= 1 and lua.isTable(1)) {
        _ = lua.getField(1, "time_budget_ms");
        if (!lua.isNil(-1)) {
            config.time_budget_ns = @intFromFloat((lua.toNumber(-1) catch 0.0) * 1_000_000.0);
        }
        lua.pop(1);

        _ = lua.getField(1, "max_iterations");
        if (!lua.isNil(-1)) {
            config.max_iterations = @intFromFloat(lua.toNumber(-1) catch 0.0);
        }
        lua.pop(1);
    }

    const allocator = std.heap.page_allocator;
    var bench = zbench.Benchmark.init(allocator, config);

    bench.add("__baseline__", struct {
        fn run(_: std.mem.Allocator) void {
            std.mem.doNotOptimizeAway(@as(u64, 0));
        }
    }.run, .{}) catch {
        _ = lua.pushString("failed to add baseline benchmark");
        return 0;
    };

    var threaded: std.Io.Threaded = .init_single_threaded;
    const io = threaded.io();

    var iter = bench.iterator() catch {
        bench.deinit();
        _ = lua.pushString("failed to create iterator");
        return 0;
    };

    var result: ?zbench.Result = null;
    while (iter.next(io) catch {
        bench.deinit();
        _ = lua.pushString("baseline iteration failed");
        return 0;
    }) |step| {
        switch (step) {
            .progress => {},
            .result => |r| {
                result = r;
            },
        }
    }

    const bench_result = result orelse {
        bench.deinit();
        _ = lua.pushString("no baseline result");
        return 0;
    };

    const timings_ns = bench_result.readings.timings_ns;
    const stats = zbench.statistics.Statistics(u64).init(timings_ns) catch {
        bench_result.deinit();
        bench.deinit();
        _ = lua.pushString("failed to compute baseline statistics");
        return 0;
    };

    lua.newTable();

    _ = lua.pushString("name");
    _ = lua.pushString("__baseline__");
    lua.setTable(-3);

    _ = lua.pushString("iterations");
    lua.pushNumber(@as(f64, @floatFromInt(bench_result.readings.iterations)));
    lua.setTable(-3);

    _ = lua.pushString("mean_ns");
    lua.pushNumber(@as(f64, @floatFromInt(stats.mean)));
    lua.setTable(-3);

    _ = lua.pushString("stddev_ns");
    lua.pushNumber(@as(f64, @floatFromInt(stats.stddev)));
    lua.setTable(-3);

    _ = lua.pushString("min_ns");
    lua.pushNumber(@as(f64, @floatFromInt(stats.min)));
    lua.setTable(-3);

    _ = lua.pushString("max_ns");
    lua.pushNumber(@as(f64, @floatFromInt(stats.max)));
    lua.setTable(-3);

    _ = lua.pushString("total_ns");
    lua.pushNumber(@as(f64, @floatFromInt(stats.total)));
    lua.setTable(-3);

    _ = lua.pushString("p75_ns");
    lua.pushNumber(@as(f64, @floatFromInt(stats.percentiles.p75)));
    lua.setTable(-3);

    _ = lua.pushString("p99_ns");
    lua.pushNumber(@as(f64, @floatFromInt(stats.percentiles.p99)));
    lua.setTable(-3);

    _ = lua.pushString("p99_9_ns");
    lua.pushNumber(@as(f64, @floatFromInt(stats.percentiles.p995)));
    lua.setTable(-3);

    bench_result.deinit();
    bench.deinit();

    return 1;
}

fn registerModule(lua: *Lua) void {
    lua.newTable();

    _ = lua.pushString("gettime");
    lua.pushFunction(zlua.wrap(luaHighResTime));
    lua.setTable(-3);

    _ = lua.pushString("run");
    lua.pushFunction(zlua.wrap(luaBenchRun));
    lua.setTable(-3);

    _ = lua.pushString("baseline");
    lua.pushFunction(zlua.wrap(luaBaseline));
    lua.setTable(-3);
}

/// Lua entry point: called by `require("zbench")`.
/// Uses `callconv(.c)` for C ABI compatibility.
/// On Windows, Zig's `export` keyword automatically handles
/// `__declspec(dllexport)` for DLL symbol visibility.
pub export fn luaopen_zbench(state: ?*zlua.LuaState) callconv(.c) c_int {
    const lua = @as(*Lua, @ptrCast(state.?));
    registerModule(lua);
    return 1;
}

// ============================================================
// Tests
// ============================================================

test "high_res_time returns valid timestamp" {
    var threaded: std.Io.Threaded = .init_single_threaded;
    const io = threaded.io();
    const t1 = std.Io.Timestamp.now(io, .awake).nanoseconds;
    const seconds1 = @as(f64, @floatFromInt(t1)) / 1_000_000_000.0;

    var i: u64 = 0;
    while (i < 100_000) : (i += 1) {
        std.mem.doNotOptimizeAway(i);
    }

    const t2 = std.Io.Timestamp.now(io, .awake).nanoseconds;
    const seconds2 = @as(f64, @floatFromInt(t2)) / 1_000_000_000.0;

    try std.testing.expect(seconds2 > seconds1);
    try std.testing.expect(seconds2 - seconds1 < 1.0);
}

test "benchmark empty function" {
    const allocator = std.testing.allocator;

    var bench = zbench.Benchmark.init(allocator, .{
        .time_budget_ns = 10_000_000,
        .max_iterations = 10,
    });
    defer bench.deinit();

    try bench.add("test_empty", struct {
        fn run(_: std.mem.Allocator) void {
            std.mem.doNotOptimizeAway(@as(u64, 0));
        }
    }.run, .{});

    var threaded: std.Io.Threaded = .init_single_threaded;
    const io = threaded.io();
    // Use iterator API to avoid outputting benchmark table to stdout
    // (stdout is used by test runner protocol in --listen mode)
    var iter = try bench.iterator();
    var result_count: usize = 0;
    while (try iter.next(io)) |step| {
        switch (step) {
            .progress => {},
            .result => |r| {
                result_count += 1;
                try std.testing.expectEqualStrings("test_empty", r.name);
                r.deinit();
            },
        }
    }
    try std.testing.expectEqual(@as(usize, 1), result_count);
}

test "benchmark with simple computation" {
    const allocator = std.testing.allocator;

    var bench = zbench.Benchmark.init(allocator, .{
        .time_budget_ns = 10_000_000,
        .max_iterations = 10,
    });
    defer bench.deinit();

    try bench.add("test_compute", struct {
        fn run(_: std.mem.Allocator) void {
            var sum: u64 = 0;
            for (0..100) |i| {
                sum += i * i;
            }
            std.mem.doNotOptimizeAway(sum);
        }
    }.run, .{});

    var threaded: std.Io.Threaded = .init_single_threaded;
    const io = threaded.io();
    // Use iterator API to avoid outputting benchmark table to stdout
    var iter = try bench.iterator();
    var result_count: usize = 0;
    while (try iter.next(io)) |step| {
        switch (step) {
            .progress => {},
            .result => |r| {
                result_count += 1;
                try std.testing.expectEqualStrings("test_compute", r.name);
                r.deinit();
            },
        }
    }
    try std.testing.expectEqual(@as(usize, 1), result_count);
}

File: src/lua-zbench.lua
-- lua-zbench: Lua microbenchmarking framework
-- Pure Lua wrapper providing a declarative DSL on top of the Zig shared library.
--
-- Usage:
--   local bench = require("lua-zbench")
--   bench.describe("My Benchmark Suite", function()
--       bench.it("fast function", function()
--           -- code to benchmark
--       end, { time_budget_ms = 500 })
--   end)
--   bench.run()
--
-- Or use the low-level API directly:
--   local results = bench.run_single("name", func, opts)

local M = {}

-- Internal state for the declarative DSL
local suites = {}       -- list of { name, benchmarks: { name, func, opts }[] }
local current_suite = nil

-- ANSI color codes for terminal output
local COLORS = {
    reset     = "\27[0m",
    bold      = "\27[1m",
    red       = "\27[31m",
    green     = "\27[32m",
    yellow    = "\27[33m",
    blue      = "\27[34m",
    magenta   = "\27[35m",
    cyan      = "\27[36m",
    white     = "\27[37m",
    gray      = "\27[90m",
}

local function has_color()
    -- Check if stdout is a terminal
    if type(io.stdout) == "userdata" and io.stdout.setvbuf then
        local ok, err = pcall(io.stdout.isatty, io.stdout)
        if ok then return err end
    end
    -- Fallback: check common environment variables
    local term = os.getenv("TERM")
    if term then
        return term ~= "dumb" and term ~= ""
    end
    return false
end

local use_color = has_color()

local function color(text, c)
    if use_color then
        return COLORS[c] .. text .. COLORS.reset
    end
    return text
end

local function bold(text)
    if use_color then
        return COLORS.bold .. text .. COLORS.reset
    end
    return text
end

-- Format nanoseconds into a human-readable string
local function format_duration(ns)
    if ns < 1000 then
        return string.format("%.2f ns", ns)
    elseif ns < 1000000 then
        return string.format("%.3f µs", ns / 1000)
    elseif ns < 1000000000 then
        return string.format("%.3f ms", ns / 1000000)
    else
        return string.format("%.3f s", ns / 1000000000)
    end
end

-- Format bytes into a human-readable string
local function format_bytes(bytes)
    if bytes < 1024 then
        return string.format("%d B", bytes)
    elseif bytes < 1024 * 1024 then
        return string.format("%.2f KiB", bytes / 1024)
    else
        return string.format("%.2f MiB", bytes / (1024 * 1024))
    end
end

-- Try to load the native zbench module
local ok, native = pcall(function()
    return require("zbench")
end)

if not ok then
    error("Failed to load native zbench module: " .. tostring(native) .. "\n" ..
          "Make sure libzbench is installed and accessible via package.cpath.")
end

-- Low-level API: run a single benchmark
-- Returns a table with: name, iterations, mean_ns, min_ns, max_ns, stddev_ns,
--                       p75_ns, p99_ns, p99_9_ns, total_ns
--                       (and optionally alloc_max_bytes, alloc_mean_bytes, alloc_count)
function M.run_single(name, func, opts)
    opts = opts or {}
    return native.run(name, func, opts)
end

-- High-resolution time function
function M.gettime()
    return native.gettime()
end

-- Declarative DSL: describe a benchmark suite
function M.describe(name, fn)
    current_suite = {
        name = name,
        benchmarks = {},
    }
    table.insert(suites, current_suite)
    fn()
    current_suite = nil
end

-- Declarative DSL: define a benchmark case
function M.it(name, func, opts)
    if not current_suite then
        error("bench.it() must be called inside a bench.describe() block")
    end
    table.insert(current_suite.benchmarks, {
        name = name,
        func = func,
        opts = opts or {},
    })
end

-- Run all registered benchmarks and print results
function M.run(opts)
    opts = opts or {}
    local json_output = opts.json or false
    local json_filename = opts.json_file or "results.json"
    local results = {}

    -- Print header
    local total_benchmarks = 0
    for _, suite in ipairs(suites) do
        total_benchmarks = total_benchmarks + #suite.benchmarks
    end

    if total_benchmarks == 0 then
        print(color("No benchmarks defined. Use bench.describe() and bench.it() to define them.", "yellow"))
        return {}
    end

    print("")
    print(bold(color("lua-zbench - Microbenchmark Results", "cyan")))
    print(color(string.rep("─", 80), "gray"))
    print("")

    for _, suite in ipairs(suites) do
        print(bold(color("Suite: " .. suite.name, "magenta")))
        print("")

        local suite_results = {}

        for _, b in ipairs(suite.benchmarks) do
            -- Run the benchmark
            local ok, result = pcall(M.run_single, b.name, b.func, b.opts)
            if not ok then
                print(color("  ✗ " .. b.name .. ": " .. tostring(result), "red"))
                table.insert(suite_results, { name = b.name, error = tostring(result) })
            else
                table.insert(suite_results, result)
                print(color("  ✓ " .. b.name, "green"))
                print(string.format("      Iterations: %s", color(tostring(result.iterations), "cyan")))
                print(string.format("      Mean:       %s", color(format_duration(result.mean_ns), "yellow")))
                print(string.format("      StdDev:     %s", color(format_duration(result.stddev_ns), "gray")))
                print(string.format("      Min:        %s", color(format_duration(result.min_ns), "green")))
                print(string.format("      Max:        %s", color(format_duration(result.max_ns), "red")))
                print(string.format("      p75:        %s", color(format_duration(result.p75_ns), "cyan")))
                print(string.format("      p99:        %s", color(format_duration(result.p99_ns), "cyan")))
                print(string.format("      p99.9:      %s", color(format_duration(result.p99_9_ns), "cyan")))

                if result.alloc_max_bytes then
                    print(string.format("      Alloc max:  %s", color(format_bytes(result.alloc_max_bytes), "blue")))
                    print(string.format("      Alloc mean: %s", color(format_bytes(result.alloc_mean_bytes), "blue")))
                    print(string.format("      Alloc cnt:  %s", color(tostring(result.alloc_count), "blue")))
                end
                print("")
            end
        end

        table.insert(results, {
            suite = suite.name,
            results = suite_results,
        })
    end

    print(color(string.rep("─", 80), "gray"))
    print(string.format("Total: %d benchmark(s) in %d suite(s)", total_benchmarks, #suites))
    print("")

    -- JSON output
    if json_output then
        local json = M.to_json(results)
        local file, err = io.open(json_filename, "w")
        if file then
            file:write(json)
            file:close()
            print(color("Results written to " .. json_filename, "green"))
        else
            print(color("Failed to write " .. json_filename .. ": " .. tostring(err), "red"))
        end
    end

    return results
end

-- Simple JSON encoder for strings
local function json_encode(s)
    if type(s) == "string" then
        -- Escape special characters
        s = string.gsub(s, '\\', '\\\\')
        s = string.gsub(s, '"', '\\"')
        s = string.gsub(s, '\n', '\\n')
        s = string.gsub(s, '\r', '\\r')
        s = string.gsub(s, '\t', '\\t')
        return '"' .. s .. '"'
    end
    return tostring(s)
end

-- Convert results to JSON string
function M.to_json(results)
    local parts = {}
    table.insert(parts, '{\n  "benchmarks": [\n')

    for i, suite in ipairs(results) do
        table.insert(parts, '    {\n')
        table.insert(parts, '      "suite": ' .. json_encode(suite.suite) .. ',\n')
        table.insert(parts, '      "results": [\n')

        for j, r in ipairs(suite.results) do
            if r.error then
                table.insert(parts, '        {\n')
                table.insert(parts, '          "name": ' .. json_encode(r.name) .. ',\n')
                table.insert(parts, '          "error": ' .. json_encode(r.error) .. '\n')
                table.insert(parts, '        }')
            else
                table.insert(parts, '        {\n')
                table.insert(parts, '          "name": ' .. json_encode(r.name) .. ',\n')
                table.insert(parts, '          "iterations": ' .. r.iterations .. ',\n')
                table.insert(parts, '          "mean_ns": ' .. r.mean_ns .. ',\n')
                table.insert(parts, '          "stddev_ns": ' .. r.stddev_ns .. ',\n')
                table.insert(parts, '          "min_ns": ' .. r.min_ns .. ',\n')
                table.insert(parts, '          "max_ns": ' .. r.max_ns .. ',\n')
                table.insert(parts, '          "total_ns": ' .. r.total_ns .. ',\n')
                table.insert(parts, '          "p75_ns": ' .. r.p75_ns .. ',\n')
                table.insert(parts, '          "p99_ns": ' .. r.p99_ns .. ',\n')
                table.insert(parts, '          "p99_9_ns": ' .. r.p99_9_ns)

                if r.alloc_max_bytes then
                    table.insert(parts, ',\n')
                    table.insert(parts, '          "alloc_max_bytes": ' .. r.alloc_max_bytes .. ',\n')
                    table.insert(parts, '          "alloc_mean_bytes": ' .. r.alloc_mean_bytes .. ',\n')
                    table.insert(parts, '          "alloc_count": ' .. r.alloc_count)
                end

                table.insert(parts, '\n        }')
            end

            if j < #suite.results then
                table.insert(parts, ',')
            end
            table.insert(parts, '\n')
        end

        table.insert(parts, '      ]\n')
        table.insert(parts, '    }')

        if i < #results then
            table.insert(parts, ',')
        end
        table.insert(parts, '\n')
    end

    table.insert(parts, '  ]\n')
    table.insert(parts, '}\n')

    return table.concat(parts)
end

-- Simple JSON encoder for strings
local function json_encode(s)
    if type(s) == "string" then
        -- Escape special characters
        s = string.gsub(s, '\\', '\\\\')
        s = string.gsub(s, '"', '\\"')
        s = string.gsub(s, '\n', '\\n')
        s = string.gsub(s, '\r', '\\r')
        s = string.gsub(s, '\t', '\\t')
        return '"' .. s .. '"'
    end
    return tostring(s)
end

-- Reset all registered suites (useful for re-running)
function M.reset()
    suites = {}
    current_suite = nil
end

-- Return the native module for advanced usage
function M.native()
    return native
end

return M
